tox.ini

```diff
@@ -10,14 +10,11 @@ deps =
     test: pytest-cov
     test: pytest-mpi>=0.2
 
-    py36-deps: cython>=0.23
-    py37-deps: cython>=0.29
```

Comment:
Should Cython still be encoded anywhere in the tox config for the `deps` builds? Or is it sufficient to have it as a dependency in the packaging info?

Reply:
We used to have a `nodeps` tox env for testing without having numpy/cython preinstalled, not sure why that was dropped (possibly it was making the CI more flakey?)

Reply:
Ah, so that might wind up using the dicey setuptools method for installing build dependencies as eggs. @tacaswell and I were talking yesterday about using PEP-518, which should avoid setuptools complicating things.

Reply:
I think it's the optional build deps (i.e. mpi4py) that cause the problem, I don't think we can do some trick with extras or something to get it to work (based on my knowledge of PEP-518).

Reply:
Aha, right, I'd forgotten about that. No, I can't see any way of expressing that at present.

Reply:
I went to post a question about this on Discourse, but found a recent thread about it, and actually it should be possible, as [Paul Ganssle explains](https://discuss.python.org/t/options-to-build-the-same-package-in-different-ways-with-different-build-dependencies/4458/10?u=takluyver).

PEP 518 specifies how to express fixed dependencies to run the build system at all (like setuptools). Then PEP 517 defines how to call the build system, including how to ask it for extra dependencies, which can depend on environment variables or config files. So we'd conditionally add mpi4py to `setup_requires`. Apparently, when setuptools is used as a PEP 517 backend, it doesn't try to install `setup_requires` as eggs.

Reply:
Oooh, that seems smart (and that could open up options for being able to generate correct dependency metadata based on what was used to build, rather than trying to lock numpy/cython down statically via pyproject.toml - have a release_deps option which uses minimum values, but also allow for building against the latest version to check how newer versions interact).